### PR TITLE
Switch Accessibility tests to `@axe-core/puppeteer`

### PR DESCRIPTION
**Pull request by colinrotherham**
This PR closes:

* https://github.com/alphagov/govuk-frontend/issues/1971

All components + examples are now tested in [`axe-core@4.7.0`](https://github.com/dequelabs/axe-core/releases/tag/v4.7.0) via [Puppeteer](https://pptr.dev)

The tests run locally via `npm test` or separately with:

```console
npx jest --selectProjects "Accessibility tests"
```

Colour contrast checks are built in ✅

## Known issues
This switch has flagged a few known issues which are now ignored + documented, including some tests with positive `tabindex` and "landmark-inside-landmark" issues due to preview HTML
  
* **All page content must be contained by landmarks**
  https://github.com/alphagov/govuk-frontend/issues/1604
  https://dequeuniversity.com/rules/axe/4.7/region

* **Accordion "`aria-labelledby` attribute cannot be used on a div with no valid role attribute"**
  https://github.com/alphagov/govuk-frontend/issues/2472
  https://dequeuniversity.com/rules/axe/4.7/aria-allowed-attr

* **Conditional reveal "ARIA attribute is not allowed: `aria-expanded="false"`"**
  https://github.com/alphagov/govuk-frontend/issues/979
  https://dequeuniversity.com/rules/axe/4.7/aria-allowed-attr

* **Invalid ARIA attribute value: `aria-controls="example-id"`** for missing IDs
  https://dequeuniversity.com/rules/axe/4.7/aria-valid-attr-value

* **Colour contrast for disabled form controls** 
  https://dequeuniversity.com/rules/axe/4.7/color-contrast

## Test reports
I've kept the [`jest-axe`](https://github.com/nickcolley/jest-axe#jest-axe) reporter but generated the report using [`@axe-core/puppeteer`](https://www.npmjs.com/package/@axe-core/puppeteer) instead

GitHub Actions will log failures as below:

```console
"Elements must meet minimum color contrast ratio thresholds (color-contrast)"

Fix any of the following:
  Element has insufficient color contrast of 2.03 (foreground color: #aaa9a9, background color: #eeefef, font size: 12.0pt (16px), font weight: bold). Expected contrast ratio of 4.5:1

You can find more information on this issue here:
https://dequeuniversity.com/rules/axe/4.7/color-contrast?application=axe-puppeteer
```

## Test workflow
Test progress appears on the GitHub Actions "Checks" tab (see [example **Accessibility tests**](https://github.com/alphagov/govuk-frontend/actions/runs/4732152847/jobs/8398000905))

Accessibility tests in GitHub Actions workflow

### CODE REVIEW

**Comment by colinrotherham** [comment]
We were previously on [`axe-core@4.5.1`](https://github.com/dequelabs/axe-core/releases/tag/v4.5.1) which is [bundled with `jest-axe`](https://github.com/NickColley/jest-axe/blob/main/package.json#L29)

File: app/src/views/layouts/full-width-landmarks.njk [hunk 1]
```diff
@@ -0,0 +1,10 @@
+{% extends "layout.njk" %}
+
+{% block main %}
+  <div class="govuk-width-container">
+    {% block beforeContent %}{% endblock %}
+  </div>
+  <main class="govuk-main-wrapper {{ mainClasses }}" id="main-content" role="main">
```
Comment (colinrotherham) [inline app/src/views/layouts/full-width-landmarks.njk:7]: I've created `full-width.njk` (without wrapping landmarks) alongside `full-width-landmarks.njk`

This is so Axe can pass some of the "landmark-in-landmark" violations

You'll see some Percy related diff changes as a result

**Review by 36degrees — APPROVED** [review]
This looks ace @colinrotherham, thanks! Seems like a sensible approach, and great commit structure.

I think it makes sense to keep a `accessibility.test.mjs` for each component whilst we have individual rule exceptions, but is there any other reason for doing it that way? Wondering if in the future we might move to having this in a single place, like [we have for HTML validation](https://github.com/alphagov/govuk-frontend/blob/ed3f083bf6493469f25268e509d529d4bad84f8d/src/govuk/components/components.template.test.js#L115-L132).

Can we make sure we add the accessibility tests to the required status checks for the main branch after this is merged?

**Comment by colinrotherham** [comment]
Thanks @36degrees 🙌

Ahh yep, it's to align with [**Running individual tests** in `testing-and-linting.md`](https://github.com/alphagov/govuk-frontend/blob/main/docs/releasing/testing-and-linting.md#running-individual-tests)

The entire test suite is quite slow, so if you're only working on one component you can narrow things down:

### Run Accordion tests only
Total time: 5s

```console
npx jest --watch src/govuk/components/accordion
```

You can try `npx jest --watch src/govuk/components/accordion --listFiles` to see what's matched:

```
src/govuk/components/accordion/accessibility.test.mjs
src/govuk/components/accordion/accordion.test.js
src/govuk/components/accordion/template.test.js
```

### Run Accessibility tests only
Total time: 55s

```console
npx jest --selectProjects "Accessibility tests"
```

### Run all non-build tests
Total time: 1m 20s (previously 25s)

```console
npm test
```

Running `npx jest` will include build tasks but **dist** and **package** directories need building first

The added test time was a concern but well worth it

**Comment by 36degrees** [comment]
That makes sense, thanks for explaining 👍🏻

**Comment by colinrotherham** [comment]
> Can we make sure we add the accessibility tests to the required status checks for the main branch after this is merged?

@36degrees Great idea

I've just fixed the conflicts and pushed again

Some extra comments added to [`jest.config.mjs`](https://github.com/alphagov/govuk-frontend/blob/axe-puppeteer/jest.config.mjs#L45) plus `maxWorkers: '50%'` to ensure Jest [uses the same default](https://jestjs.io/docs/cli#--maxworkersnumstring) across single project runs, multi-project runs, and with flags like `--watch` and `--watchAll`

Addresses a minor concern that 12x concurrent Chromium processes on my machine was a bit much!

```console
npx jest --selectProjects "Accessibility tests"
```

## Support for macOS ARM CPUs
Out of interest, but performance related, can you test this for me? I've got an Intel Mac

Enabling the ARM CPU experiment to see how much faster things run:

See config [**puppeteer.config.js**](https://github.com/alphagov/govuk-frontend/blob/axe-puppeteer/puppeteer.config.js)
```js
module.exports = {
  cacheDirectory: join(__dirname, '.cache', 'puppeteer'),
  experiments: {
    macArmChromiumEnabled: true
  }
}
```

Thanks

**Comment by colinrotherham** [comment]
Thought it best to push more commits to explain how to _see_ these tests in action

Since our headless browser tests can be made visible with `{ headless: false }` I've added:

1. Documentation for `await jestPuppeteer.debug()` to pause tests
2. Jest config changes to only open 1x browser window
3. Jest teardown changes to keep browsers open

**Review by 36degrees — APPROVED** [review]
The new commits look good to me, I'm just not sure about the documentation.

**Comment by colinrotherham** [comment]
I've made the suggested documentation change but with `HEADLESS=false` instead of `{ headless: false }`

Branch protection settings also have **Accessibility tests** added 👍

(Other open PRs will need updating to include the new tests)

**Pull request merged**